File: backend/app/models/persona.py
"""Persona model - user's professional identity.

REQ-005 §4.1 - Tier 1, references User.
"""

import uuid
from typing import TYPE_CHECKING

from sqlalchemy import (
    Boolean,
    CheckConstraint,
    ForeignKey,
    Integer,
    String,
    Text,
    text,
)
from sqlalchemy.dialects.postgresql import JSONB, UUID
from sqlalchemy.orm import Mapped, mapped_column, relationship

from app.models.base import Base, TimestampMixin

_CASCADE_ALL_DELETE_ORPHAN = "all, delete-orphan"

if TYPE_CHECKING:
    from app.models.application import Application
    from app.models.cover_letter import CoverLetter
    from app.models.job_posting import JobPosting
    from app.models.job_source import PollingConfiguration, UserSourcePreference
    from app.models.persona_content import (
        AchievementStory,
        Certification,
        Education,
        Skill,
        WorkHistory,
    )
    from app.models.persona_settings import (
        CustomNonNegotiable,
        PersonaChangeFlag,
        PersonaEmbedding,
        VoiceProfile,
    )
    from app.models.resume import BaseResume, ResumeFile
    from app.models.user import User


_DEFAULT_EMPTY_JSONB = text("'[]'::jsonb")


class Persona(Base, TimestampMixin):
    """User's professional identity containing all career data.

    Central entity linking to work history, skills, education, etc.
    Tier 1 - references User.
    """

    __tablename__ = "personas"

    # Primary key
    id: Mapped[uuid.UUID] = mapped_column(
        UUID(as_uuid=True),
        primary_key=True,
        server_default=text("gen_random_uuid()"),
    )

    # User reference (Tier 1 dependency)
    user_id: Mapped[uuid.UUID] = mapped_column(
        UUID(as_uuid=True),
        ForeignKey("users.id", ondelete="CASCADE"),
        nullable=False,
    )

    # Contact information
    email: Mapped[str] = mapped_column(
        String(255),
        unique=True,
        nullable=False,
    )
    full_name: Mapped[str] = mapped_column(
        String(255),
        nullable=False,
    )
    phone: Mapped[str] = mapped_column(
        String(50),
        nullable=False,
    )

    # Location
    home_city: Mapped[str] = mapped_column(
        String(100),
        nullable=False,
    )
    home_state: Mapped[str] = mapped_column(
        String(100),
        nullable=False,
    )
    home_country: Mapped[str] = mapped_column(
        String(100),
        nullable=False,
    )

    # Online presence
    linkedin_url: Mapped[str | None] = mapped_column(
        String(500),
        nullable=True,
    )
    portfolio_url: Mapped[str | None] = mapped_column(
        String(500),
        nullable=True,
    )

    # Professional summary
    professional_summary: Mapped[str | None] = mapped_column(
        Text,
        nullable=True,
    )
    years_experience: Mapped[int | None] = mapped_column(
        Integer,
        nullable=True,
    )
    current_role: Mapped[str | None] = mapped_column(
        String(255),
        nullable=True,
    )
    current_company: Mapped[str | None] = mapped_column(
        String(255),
        nullable=True,
    )

    # Career goals (JSONB arrays)
    target_roles: Mapped[list] = mapped_column(
        JSONB,
        server_default=_DEFAULT_EMPTY_JSONB,
        nullable=False,
    )
    target_skills: Mapped[list] = mapped_column(
        JSONB,
        server_default=_DEFAULT_EMPTY_JSONB,
        nullable=False,
    )

    # Location preferences (JSONB arrays)
    commutable_cities: Mapped[list] = mapped_column(
        JSONB,
        server_default=_DEFAULT_EMPTY_JSONB,
        nullable=False,
    )
    relocation_cities: Mapped[list] = mapped_column(
        JSONB,
        server_default=_DEFAULT_EMPTY_JSONB,
        nullable=False,
    )
    industry_exclusions: Mapped[list] = mapped_column(
        JSONB,
        server_default=_DEFAULT_EMPTY_JSONB,
        nullable=False,
    )

    # Preferences
    stretch_appetite: Mapped[str] = mapped_column(
        String(20),
        server_default=text("'Medium'"),
        nullable=False,
    )
    minimum_base_salary: Mapped[int | None] = mapped_column(
        Integer,
        nullable=True,
    )
    salary_currency: Mapped[str] = mapped_column(
        String(10),
        server_default=text("'USD'"),
        nullable=False,
    )
    max_commute_minutes: Mapped[int | None] = mapped_column(
        Integer,
        nullable=True,
    )
    remote_preference: Mapped[str] = mapped_column(
        String(30),
        server_default=text("'No Preference'"),
        nullable=False,
    )
    relocation_open: Mapped[bool] = mapped_column(
        Boolean,
        server_default=text("false"),
        nullable=False,
    )
    visa_sponsorship_required: Mapped[bool] = mapped_column(
        Boolean,
        server_default=text("false"),
        nullable=False,
    )
    company_size_preference: Mapped[str] = mapped_column(
        String(30),
        server_default=text("'No Preference'"),
        nullable=False,
    )
    max_travel_percent: Mapped[str] = mapped_column(
        String(20),
        server_default=text("'Any'"),
        nullable=False,
    )

    # Matching thresholds
    minimum_fit_threshold: Mapped[int] = mapped_column(
        Integer,
        server_default=text("50"),
        nullable=False,
    )
    auto_draft_threshold: Mapped[int] = mapped_column(
        Integer,
        server_default=text("90"),
        nullable=False,
    )

    # Polling settings
    polling_frequency: Mapped[str] = mapped_column(
        String(20),
        server_default=text("'Daily'"),
        nullable=False,
    )

    # Onboarding state
    onboarding_complete: Mapped[bool] = mapped_column(
        Boolean,
        server_default=text("false"),
        nullable=False,
    )
    onboarding_step: Mapped[str | None] = mapped_column(
        String(50),
        nullable=True,
    )

    # Original uploaded resume (nullable FK added in Tier 2)
    # use_alter=True handles circular dependency with ResumeFile at migration level
    original_resume_file_id: Mapped[uuid.UUID | None] = mapped_column(
        UUID(as_uuid=True),
        ForeignKey(
            "resume_files.id",
            ondelete="SET NULL",
            use_alter=True,
            name="fk_persona_original_resume_file",
        ),
        nullable=True,
    )

    __table_args__ = (
        CheckConstraint(
            "stretch_appetite IN ('Low', 'Medium', 'High')",
            name="ck_persona_stretch_appetite",
        ),
        CheckConstraint(
            "remote_preference IN ('Remote Only', 'Hybrid OK', 'Onsite OK', 'No Preference')",
            name="ck_persona_remote_preference",
        ),
        CheckConstraint(
            "company_size_preference IN ('Startup', 'Mid-size', 'Enterprise', 'No Preference')",
            name="ck_persona_company_size",
        ),
        CheckConstraint(
            "max_travel_percent IN ('None', '<25%', '<50%', 'Any')",
            name="ck_persona_max_travel",
        ),
        CheckConstraint(
            "polling_frequency IN ('Daily', 'Twice Daily', 'Weekly', 'Manual Only')",
            name="ck_persona_polling_frequency",
        ),
        CheckConstraint(
            "minimum_fit_threshold >= 0 AND minimum_fit_threshold <= 100",
            name="ck_persona_fit_threshold",
        ),
        CheckConstraint(
            "auto_draft_threshold >= 0 AND auto_draft_threshold <= 100",
            name="ck_persona_draft_threshold",
        ),
    )

    # Relationships to User (Tier 0)
    user: Mapped["User"] = relationship(
        "User",
        back_populates="personas",
    )

    # Relationships to Tier 2 persona content
    work_histories: Mapped[list["WorkHistory"]] = relationship(
        "WorkHistory",
        back_populates="persona",
        cascade=_CASCADE_ALL_DELETE_ORPHAN,
    )
    skills: Mapped[list["Skill"]] = relationship(
        "Skill",
        back_populates="persona",
        cascade=_CASCADE_ALL_DELETE_ORPHAN,
    )
    educations: Mapped[list["Education"]] = relationship(
        "Education",
        back_populates="persona",
        cascade=_CASCADE_ALL_DELETE_ORPHAN,
    )
    certifications: Mapped[list["Certification"]] = relationship(
        "Certification",
        back_populates="persona",
        cascade=_CASCADE_ALL_DELETE_ORPHAN,
    )
    achievement_stories: Mapped[list["AchievementStory"]] = relationship(
        "AchievementStory",
        back_populates="persona",
        cascade=_CASCADE_ALL_DELETE_ORPHAN,
    )

    # Relationships to Tier 2 persona settings
    voice_profile: Mapped["VoiceProfile | None"] = relationship(
        "VoiceProfile",
        back_populates="persona",
        uselist=False,
        cascade=_CASCADE_ALL_DELETE_ORPHAN,
    )
    custom_non_negotiables: Mapped[list["CustomNonNegotiable"]] = relationship(
        "CustomNonNegotiable",
        back_populates="persona",
        cascade=_CASCADE_ALL_DELETE_ORPHAN,
    )
    embeddings: Mapped[list["PersonaEmbedding"]] = relationship(
        "PersonaEmbedding",
        back_populates="persona",
        cascade=_CASCADE_ALL_DELETE_ORPHAN,
    )
    change_flags: Mapped[list["PersonaChangeFlag"]] = relationship(
        "PersonaChangeFlag",
        back_populates="persona",
        cascade=_CASCADE_ALL_DELETE_ORPHAN,
    )

    # Relationships to resume domain
    resume_files: Mapped[list["ResumeFile"]] = relationship(
        "ResumeFile",
        back_populates="persona",
        cascade=_CASCADE_ALL_DELETE_ORPHAN,
        foreign_keys="ResumeFile.persona_id",
    )
    original_resume_file: Mapped["ResumeFile | None"] = relationship(
        "ResumeFile",
        foreign_keys=[original_resume_file_id],
        post_update=True,
    )
    base_resumes: Mapped[list["BaseResume"]] = relationship(
        "BaseResume",
        back_populates="persona",
        cascade=_CASCADE_ALL_DELETE_ORPHAN,
    )

    # Relationships to job source preferences
    source_preferences: Mapped[list["UserSourcePreference"]] = relationship(
        "UserSourcePreference",
        back_populates="persona",
        cascade=_CASCADE_ALL_DELETE_ORPHAN,
    )
    polling_configuration: Mapped["PollingConfiguration | None"] = relationship(
        "PollingConfiguration",
        back_populates="persona",
        uselist=False,
        cascade=_CASCADE_ALL_DELETE_ORPHAN,
    )

    # Relationships to job/application domain
    job_postings: Mapped[list["JobPosting"]] = relationship(
        "JobPosting",
        back_populates="persona",
        cascade=_CASCADE_ALL_DELETE_ORPHAN,
    )
    cover_letters: Mapped[list["CoverLetter"]] = relationship(
        "CoverLetter",
        back_populates="persona",
        cascade=_CASCADE_ALL_DELETE_ORPHAN,
    )
    applications: Mapped[list["Application"]] = relationship(
        "Application",
        back_populates="persona",
        cascade=_CASCADE_ALL_DELETE_ORPHAN,
    )

File: frontend/src/types/persona.ts
/**
 * Persona domain types matching backend/app/models/persona*.py.
 *
 * REQ-001: Professional persona data model.
 * REQ-005 §4.1: Database schema (Tier 1 Persona, Tier 2 sub-entities).
 * REQ-012 §4: Frontend architecture.
 */

// ---------------------------------------------------------------------------
// Enum union types — match backend CHECK constraints
// ---------------------------------------------------------------------------

/** Backend: WorkHistory.work_model CHECK constraint. */
export type WorkModel = "Remote" | "Hybrid" | "Onsite";

/** Backend: Skill.skill_type CHECK constraint. */
export type SkillType = "Hard" | "Soft";

/** Backend: Skill.proficiency CHECK constraint. Ordered ascending. */
export type Proficiency = "Learning" | "Familiar" | "Proficient" | "Expert";

/** Backend: Persona.remote_preference CHECK constraint. */
export type RemotePreference =
	| "Remote Only"
	| "Hybrid OK"
	| "Onsite OK"
	| "No Preference";

/** Backend: Persona.company_size_preference CHECK constraint. */
export type CompanySizePreference =
	| "Startup"
	| "Mid-size"
	| "Enterprise"
	| "No Preference";

/** Backend: Persona.max_travel_percent CHECK constraint. */
export type MaxTravelPercent = "None" | "<25%" | "<50%" | "Any";

/** Backend: Persona.stretch_appetite CHECK constraint. */
export type StretchAppetite = "Low" | "Medium" | "High";

/** Backend: Persona.polling_frequency CHECK constraint. */
export type PollingFrequency =
	| "Daily"
	| "Twice Daily"
	| "Weekly"
	| "Manual Only";

/** Backend: CustomNonNegotiable.filter_type CHECK constraint. */
export type FilterType = "Exclude" | "Require";

/** Backend: PersonaChangeFlag.change_type CHECK constraint. */
export type ChangeType =
	| "job_added"
	| "bullet_added"
	| "skill_added"
	| "education_added"
	| "certification_added";

/** Backend: PersonaChangeFlag.status CHECK constraint. */
export type ChangeFlagStatus = "Pending" | "Resolved";

/** Backend: PersonaChangeFlag.resolution CHECK constraint. */
export type ChangeFlagResolution = "added_to_all" | "added_to_some" | "skipped";

// ---------------------------------------------------------------------------
// Enum value arrays — for form dropdowns and validation
// ---------------------------------------------------------------------------

export const WORK_MODELS: readonly WorkModel[] = [
	"Remote",
	"Hybrid",
	"Onsite",
] as const;

export const PROFICIENCIES: readonly Proficiency[] = [
	"Learning",
	"Familiar",
	"Proficient",
	"Expert",
] as const;

export const REMOTE_PREFERENCES: readonly RemotePreference[] = [
	"Remote Only",
	"Hybrid OK",
	"Onsite OK",
	"No Preference",
] as const;

export const COMPANY_SIZE_PREFERENCES: readonly CompanySizePreference[] = [
	"Startup",
	"Mid-size",
	"Enterprise",
	"No Preference",
] as const;

export const MAX_TRAVEL_PERCENTS: readonly MaxTravelPercent[] = [
	"None",
	"<25%",
	"<50%",
	"Any",
] as const;

export const STRETCH_APPETITES: readonly StretchAppetite[] = [
	"Low",
	"Medium",
	"High",
] as const;

export const POLLING_FREQUENCIES: readonly PollingFrequency[] = [
	"Daily",
	"Twice Daily",
	"Weekly",
	"Manual Only",
] as const;

export const FILTER_TYPES: readonly FilterType[] = [
	"Exclude",
	"Require",
] as const;

export const CHANGE_TYPES: readonly ChangeType[] = [
	"job_added",
	"bullet_added",
	"skill_added",
	"education_added",
	"certification_added",
] as const;

export const CHANGE_FLAG_STATUSES: readonly ChangeFlagStatus[] = [
	"Pending",
	"Resolved",
] as const;

export const CHANGE_FLAG_RESOLUTIONS: readonly ChangeFlagResolution[] = [
	"added_to_all",
	"added_to_some",
	"skipped",
] as const;

// ---------------------------------------------------------------------------
// Sub-entity interfaces — Tier 2 & 3 models
// ---------------------------------------------------------------------------

/**
 * Accomplishment bullet within a work history entry.
 *
 * Backend: Bullet model (persona_content.py). Tier 3 — references WorkHistory.
 */
export interface Bullet {
	id: string;
	work_history_id: string;
	text: string;
	/** UUID references to Skill entities. */
	skills_demonstrated: string[];
	metrics: string | null;
	display_order: number;
}

/**
 * Work history entry with nested accomplishment bullets.
 *
 * Backend: WorkHistory model (persona_content.py). Tier 2 — references Persona.
 */
export interface WorkHistory {
	id: string;
	persona_id: string;
	company_name: string;
	company_industry: string | null;
	job_title: string;
	/** ISO date string (YYYY-MM-DD). */
	start_date: string;
	/** ISO date string. Null when is_current is true. */
	end_date: string | null;
	is_current: boolean;
	location: string;
	work_model: WorkModel;
	description: string | null;
	display_order: number;
	bullets: Bullet[];
}

/**
 * Skill entry — hard or soft.
 *
 * Backend: Skill model (persona_content.py). Tier 2 — references Persona.
 * Unique constraint: (persona_id, skill_name).
 */
export interface Skill {
	id: string;
	persona_id: string;
	skill_name: string;
	skill_type: SkillType;
	category: string;
	proficiency: Proficiency;
	years_used: number;
	/** "Current" or a year string (e.g., "2023"). */
	last_used: string;
	display_order: number;
}

/**
 * Education entry.
 *
 * Backend: Education model (persona_content.py). Tier 2 — references Persona.
 */
export interface Education {
	id: string;
	persona_id: string;
	institution: string;
	degree: string;
	field_of_study: string;
	graduation_year: number;
	/** Decimal GPA (0.00–4.00). Null if not provided. */
	gpa: number | null;
	honors: string | null;
	display_order: number;
}

/**
 * Professional certification.
 *
 * Backend: Certification model (persona_content.py). Tier 2 — references Persona.
 */
export interface Certification {
	id: string;
	persona_id: string;
	certification_name: string;
	issuing_organization: string;
	/** ISO date string. */
	date_obtained: string;
	/** ISO date string. Null means no expiration. */
	expiration_date: string | null;
	credential_id: string | null;
	verification_url: string | null;
	display_order: number;
}

/**
 * Achievement story in Context–Action–Outcome (CAO/STAR) format.
 *
 * Backend: AchievementStory model (persona_content.py). Tier 2 — references Persona.
 */
export interface AchievementStory {
	id: string;
	persona_id: string;
	title: string;
	/** 1-2 sentences describing the situation. */
	context: string;
	action: string;
	outcome: string;
	/** UUID references to Skill entities. */
	skills_demonstrated: string[];
	/** UUID reference to WorkHistory. Null if unlinked. */
	related_job_id: string | null;
	display_order: number;
}

/**
 * Writing voice preferences for content generation.
 *
 * Backend: VoiceProfile model (persona_settings.py). One-to-one with Persona.
 */
export interface VoiceProfile {
	id: string;
	persona_id: string;
	tone: string;
	sentence_style: string;
	vocabulary_level: string;
	personality_markers: string | null;
	sample_phrases: string[];
	/** Words/phrases to avoid in generated content. */
	things_to_avoid: string[];
	writing_sample_text: string | null;
	created_at: string;
	updated_at: string;
}

/**
 * User-defined filter rule for job matching.
 *
 * Backend: CustomNonNegotiable model (persona_settings.py). Tier 2 — references Persona.
 */
export interface CustomNonNegotiable {
	id: string;
	persona_id: string;
	filter_name: string;
	filter_type: FilterType;
	filter_value: string;
	/** Job field to check (e.g., "company_name", "description", "job_title"). */
	filter_field: string;
}

/**
 * Tracks persona changes needing HITL review for base resume updates.
 *
 * Backend: PersonaChangeFlag model (persona_settings.py). Tier 2 — references Persona.
 */
export interface PersonaChangeFlag {
	id: string;
	persona_id: string;
	change_type: ChangeType;
	item_id: string;
	item_description: string;
	status: ChangeFlagStatus;
	resolution: ChangeFlagResolution | null;
	/** ISO 8601 datetime. Null when status is Pending. */
	resolved_at: string | null;
	created_at: string;
}

// ---------------------------------------------------------------------------
// Main Persona interface — Tier 1 model
// ---------------------------------------------------------------------------

/**
 * User's professional persona — the core domain entity.
 *
 * Backend: Persona model (persona.py). Tier 1 — references User.
 * Sub-entities (WorkHistory, Skills, etc.) are fetched via separate endpoints.
 */
export interface Persona {
	id: string;
	user_id: string;

	// Basic Info (REQ-001 §3.1)
	full_name: string;
	email: string;
	phone: string;
	home_city: string;
	home_state: string;
	home_country: string;
	linkedin_url: string | null;
	portfolio_url: string | null;

	// Professional Overview (REQ-001 §3.1b)
	professional_summary: string | null;
	years_experience: number | null;
	current_role: string | null;
	current_company: string | null;

	// Growth Targets (REQ-001 §3.9)
	target_roles: string[];
	target_skills: string[];
	stretch_appetite: StretchAppetite;

	// Non-Negotiables: Location (REQ-001 §3.8.1)
	commutable_cities: string[];
	max_commute_minutes: number | null;
	remote_preference: RemotePreference;
	relocation_open: boolean;
	relocation_cities: string[];

	// Non-Negotiables: Compensation (REQ-001 §3.8.2)
	minimum_base_salary: number | null;
	salary_currency: string;

	// Non-Negotiables: Other Filters (REQ-001 §3.8.3)
	visa_sponsorship_required: boolean;
	industry_exclusions: string[];
	company_size_preference: CompanySizePreference;
	max_travel_percent: MaxTravelPercent;

	// Discovery Preferences (REQ-001 §3.10)
	minimum_fit_threshold: number;
	auto_draft_threshold: number;
	polling_frequency: PollingFrequency;

	// Onboarding state
	onboarding_complete: boolean;
	onboarding_step: string | null;

	// Timestamps
	created_at: string;
	updated_at: string;
}
